RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc

```diff
@@ -0,0 +1,516 @@
+
+math::XYZPoint HGCalImagingAlgo::calculatePositionWithFraction(const std::vector<KDNode>& hits,
+								 const std::vector<double>& fractions) {  
+  double norm(0.0), x(0.0), y(0.0), z(0.0);
+  for( unsigned i = 0; i < hits.size(); ++i ) {
+    const double weight = fractions[i]*hits[i].data.weight;
+    norm += weight;
+    x += weight*hits[i].data.x;
+    y += weight*hits[i].data.y;
+    z += weight*hits[i].data.z;
+  }
+  math::XYZPoint result(x,y,z);
+  double norm_inv = 1.0/norm;
+  result *= norm_inv;
+  return result;
+}
+
+double HGCalImagingAlgo::calculateEnergyWithFraction(const std::vector<KDNode>& hits,
+						     const std::vector<double>& fractions) {
+  double result = 0.0;
+  for( unsigned i = 0 ; i < hits.size(); ++i ) {
+    result += fractions[i]*hits[i].data.weight;
+  }
+  return result;
+}
+
+void HGCalImagingAlgo::shareEnergy(const std::vector<KDNode>& incluster,
+				   const std::vector<unsigned>& seeds, 
+				   std::vector<std::vector<double> >& outclusters) {
+  std::vector<bool> isaseed(incluster.size(),false);
+  outclusters.clear();
+  outclusters.resize(seeds.size());
+  std::vector<Point> centroids(seeds.size());
+  std::vector<double> energies(seeds.size());
+
+  if( seeds.size() == 1 ) { // short circuit the case of a lone cluster
+    outclusters[0].clear();
+    outclusters[0].resize(incluster.size(),1.0);    
+    return;
+  }
+
+  //std::cout << "saving seeds" << std::endl;
+
+  // create quick seed lookup
+  for( unsigned i = 0; i < seeds.size(); ++i ) {
+    isaseed[seeds[i]] = true;
+  }
+
+  //std::cout << "seeds saved" << std::endl;
+
+  // initialize clusters to be shared
+  // centroids start off at seed positions
+  // seeds always have fraction 1.0, to stabilize fit
+  //std::cout << "initializing fit" << std::endl;
+  for( unsigned i = 0; i < seeds.size(); ++i ) {
+    outclusters[i].resize(incluster.size(),0.0);
+    for( unsigned j = 0; j < incluster.size(); ++j ) {
+      if( j == seeds[i] ) {
+	outclusters[i][j] = 1.0;
+	centroids[i] = math::XYZPoint(incluster[j].data.x,incluster[j].data.y,incluster[j].data.z);
+	energies[i]  = incluster[j].data.weight; 
+      } 
+    }
+  }
+
+  //return;
+
+  //std::cout << "fit initialized" << std::endl;
+
+  // run the fit while we are less than max iterations, and clusters are still moving
+  const double minFracTot = 1e-20;
+  unsigned iter = 0;
+  const unsigned iterMax = 50;
+  double diff = std::numeric_limits<double>::max();
+  const double stoppingTolerance = 1e-8;
+  const double toleranceScaling = std::pow(std::max(1.0,seeds.size()-1.0),2.0);
+  std::vector<Point> prevCentroids;
+  std::vector<double> frac(seeds.size()), dist2(seeds.size());
+  while( iter++ < iterMax && diff > stoppingTolerance*toleranceScaling ) {    
+    for( unsigned i = 0; i < incluster.size(); ++i ) {
+      const Hexel& ihit = incluster[i].data;
+      double fraction(0.0), fracTot(0.0), d2(0.0);
+      for( unsigned j = 0; j < seeds.size(); ++j ) {
+	fraction = 0.0;
```

Comment:
This assignment is redundant. `fraction` is certain to be assigned in lines 478-483 before it is read. But this problem is minor and could be cleaned up in a later PR.